picky-server/src/http/controller.rs

```diff
@@ -313,6 +319,82 @@ impl ServerController {
+    async fn sign_ssh_key(&self, req: Request) -> Result<ResponseBuilder, StatusCode> {
+        use serde::{Deserialize, Serialize};
+
+        #[derive(Deserialize, Serialize)]
+        struct SshSignRequest {
+            key: String,
+            key_type: SshKeyType,
+            key_id: String,
+            principals: Vec<String>,
+            duration_secs: u64,
+            nbf: u64,
+            exp: u64,
+        }
+
+        let sign_request: SshSignRequest = match check_authorization(&*self.read_conf().await, &req) {
+            Ok(token) => serde_json::from_value(token.claims).bad_request()?,
+            Err(e) => {
+                log::error!("authorization failed: {}", e);
+                return Err(StatusCode::UNAUTHORIZED);
+            }
+        };
+
+        let ssh_public_key = SshPublicKey::from_str(sign_request.key.as_str()).bad_request()?;
+
+        let sing_key = self
+            .storage
+            .get_ssh_private_key_by_type(sign_request.key_type)
+            .await
+            .internal_error()?;
+
+        let config = self.read_conf().await;
+
+        let builder = SshCertificateBuilder::init();
+
+        let (cert_type, sign_key_passphrase) = match sign_request.key_type {
+            SshKeyType::Client => (SshCertType::Client, config.ssh_client_passphrase.clone()),
+            SshKeyType::Host => (SshCertType::Host, config.ssh_host_passphrase.clone()),
+        };
+
+        let ssh_private_key = SshPrivateKey::from_pem_str(sing_key.key.as_str(), Some(sign_key_passphrase))
+            .internal_error_desc("couldn't deserialize SSH private key")?;
+        let now = OffsetDateTime::now_utc();
+        let valid_before = (now.unix_timestamp() + i64::try_from(sign_request.duration_secs).bad_request()?) as u64;
+
+        let ssh_cert = builder
+            .cert_key_type(SshCertKeyType::SshRsaV01)
+            .key(&ssh_public_key)
+            .key_id(sign_request.key_id)
+            .cert_type(cert_type)
+            .principals(sign_request.principals)
+            .valid_after(SshTime::from(now))
+            .valid_before(SshTime::from(valid_before))
+            .signature_key(&ssh_private_key)
+            .comment(ssh_public_key.comment.clone())
+            .signature_algo(config.signing_algorithm)
+            .build()
+            .internal_error()?;
+
+        let ssh_cert_pem = ssh_cert
+            .to_string()
+            .internal_error_desc("Failed to convert SSH certificate to PEM format")?;
+
+        Ok(ResponseBuilder::new().body(ssh_cert_pem).status(StatusCode::OK))
+    }
+
+    #[get("/ssh/client_ssh_key")]
+    async fn get_client_public_ssh_key(&self, _req: Request) -> Result<ResponseBuilder, StatusCode> {
+        self.get_ssh_key(SshKeyType::Client).await
+    }
+
+    #[get("/ssh/host_ssh_key")]
+    async fn get_host_public_ssh_key(&self, _req: Request) -> Result<ResponseBuilder, StatusCode> {
+        self.get_ssh_key(SshKeyType::Host).await
+    }
```

Comment:
I'm not sure we even need these routes, but if we end up keeping these, I suggest using a query parameter instead of two distinct routes (ie: `/ssh/key?type=host`).